src/frontend/src/crypto/mnemonic.ts

```diff
@@ -1,18 +1,183 @@
   crypto.getRandomValues(entropy);
-  return entropyToMnemonic(toHexString(entropy.buffer), wordlists.english);
+  return entropyToMnemonic(
+    toHexString(entropy.buffer),
+    RECOVERYPHRASE_WORDLIST
+  );
 }
 
+/** How many words are expected in the recovery phrase */
+export const RECOVERYPHRASE_WORDCOUNT = 24;
+
+/** The list of words used to make the recovery phrase */
+// NOTE: "english" is the only one actually bundled in (see webpack config)
+export const RECOVERYPHRASE_WORDLIST: string[] = wordlists.english;
+
 /**
  * @returns true if the mnemonic is valid, false otherwise.
  */
 export function validate(mnemonic: string): boolean {
   return validateMnemonic(mnemonic);
 }
+
+/** A recovery phrase that was parsed.
+ * An object of this type should not be used to reconstruct a recovery phrase, since
+ * some information (though most likely incorrectly inserted) like extra whitespaces
+ * is lost.
+ */
+type ParsedRecoveryPhrase = {
+  leadingUserNumber: string | null;
+  words: string[];
+};
+
+/** Parse recovery phrase into a structured object */
+export const parseRecoveryPhrase = (s: string): ParsedRecoveryPhrase => {
+  /* Split nicely into words
+   * ' 2002 foo  bar ' -> split(" ") -> [ '', '2002', 'foo', '', 'bar', '' ]
+   * [ '', '2002', 'foo', '', 'bar', '' ] -> filter(...) -> [ '2002', 'foo', 'bar' ]
+   */
+  const words = s.split(" ").filter((w) => w !== "");
+
+  // check if there's a leading user (anchor) number
+  const leadingUserNumber: string | null =
+    words.length >= 1 && isUserNumber(words[0]) ? words[0] : null;
+
+  // If the first element is indeed a number (BigInt), drop it from the actual mnemonic
+  if (leadingUserNumber !== null) {
+    words.shift();
+  }
+
+  return { leadingUserNumber, words };
+};
+
+/** Drop the anchor number from a string
+ * Example: " 10005 foo bar" -> "foo bar"
+ */
+export const dropLeadingUserNumber = (input: string): string => {
+  const parsed = parseRecoveryPhrase(input);
+  return parsed.leadingUserNumber !== null
+    ? input.replace(parsed.leadingUserNumber, "").trim()
+    : input;
+};
+
+/** The warning types that may be generated. For more information on each, see 'recoveryPhraseWarnings'.
+ *
+ * We generate the messages independently mostly so that we can programatically test the warnings generated
+ * (and avoid brittle regex-based testing).*/
+export type Warning =
+  | { type: "bad_chars"; chars: string[] }
+  | { type: "repeated_whitespace"; between: [string, string] | null }
+  | { type: "bad_anchor"; anchor: bigint }
+  | { type: "bad_word_count"; count: number }
+  | { type: "bad_words"; words: string[] }
+  | { type: "invalid" };
+
+/** Generates warnings for a given recovery phrase and user number */
+export const recoveryPhraseWarnings: {
+  (userNumber: bigint, inputRaw: string): Warning | null;
+}[] = [
+  // The input contains some characters that are not expected in a recovery phrase
+  (userNumber: bigint, inputRaw: string): Warning | null => {
```

Comment:
For readability I would have created each function separately and then the array:

```
export const recoveryPhraseWarnings = [validateBadChars, validateWhiteSpace, ...];
```

Reply:
thought about doing that, although that adds a few bit of (not so useful) code

Reply:
not a strong opinion, just a personal preference. But I understand your choice as well.